### ISSUE TITLE
CA: UCSD Results Staging Connectivity

### ISSUE THREAD

**Issue by scleary1cs** [CDCgov/trusted-intermediary#1031]
# Story
As UCSD, so that I can review and provide patient care based on NBS findings, I need sample result messages sent by ReportStream staging to appear in my EHR test environment.

## Pre-conditions
- Out of scope: production connectivity

## Acceptance Criteria
- [x] A sample ORU message sent by ReportStream staging to the UCSD receiver appears in their test EHR environment.
- [x] Transmission uses TLS 1.2 or greater and a strong cipher

## Tasks
- [x] Receive EHR credentials using a secure connection 
- [x] Test connection by sending messages manually
- [x] Notify UCSD that we'll be sending sample results messages
- [x] Verify receipt of messages
- [x] Configure a receiver for UCSD
- [x] Testing staging connectivity via REST
- ~~Create How to onboard document (originally in #1136 ); moving to another backlog item ~~
~~OAuth2 integration~~ Moved to #1136

## Definition of Done
- [ ] Documentation tasks completed
  - [ ] Documentation and diagrams created or updated
    - [ ] Implementation guide (`/ig` folder)
    - [ ] ADRs (`/adr` folder)
    - [ ] Main `README.md`
    - [ ] Other READMEs in the repo
    - [ ] If applicable, update the `ReportStream Setup` section in `README.md`
  - [ ] [Threat model](https://lucid.app/lucidchart/8c6e8d37-2612-42a8-ac57-150c83f8e29e/edit) updated
  - [ ] API documentation updated
- [ ] Code quality tasks completed
  - [ ] Code refactored for clarity and no design/technical debt
  - [ ] Adhere to separation of concerns; code is not tightly coupled, especially to 3rd party dependencies
  - [ ] Code is reviewed or developed by pair; 1 approval is needed but consider requiring an outside-the-pair reviewer
  - [ ] Code quality checks passed
- [ ] Security & Privacy tasks completed
  - [ ] Security & privacy gates passed
- [ ] Testing tasks completed
  - [ ] Load tests passed
  - [ ] Unit test coverage of our code >= 90%
- [ ] Build & Deploy tasks completed
  - [ ] Build process updated
  - [ ] API(s) are versioned
  - [ ] Feature toggles created and/or deleted.  Document the feature toggle
  - [ ] Source code is merged to the main branch

## Research Questions
- *Optional: Any initial questions for research*

## Decisions
- *Optional: Any decisions we've made while working on this story*

## Notes
- *Optional: Any reference material or thoughts we may need for later reference*

**Comment by JohnNKing** [comment]
Trying to figure out issues with UCSD-specific OAuth endpoint; not able to authenticate successfully.

@JohnNKing Add onboarding document task here (from OAuth story)

**Comment by jorg3lopez** [comment]
Update:
- OAuth2 [PR ](https://github.com/CDCgov/prime-reportstream/pull/14869)has been merged into RS. This gives RS the capability to create a customized jwt when using the REST transport. (Essential for CA integration)
- We are able to retrieve access token from UCSD EPIC's token endpoint.
- Connection to UCSD EPIC's environment was successfully established. Our team was able to send a payload successfully. We are looking into cleaning the return cartridge and new line from the message so it doesn't error out. Also, there is an invalid HL7 version ACK that we get back. We will look into that as well.

Successful token retrieval (expired access token):
![image](https://github.com/user-attachments/assets/9fc82099-07d8-4fcf-a1b1-ddb562f585fb)

Invalid HL7 version NACK:
![image](https://github.com/user-attachments/assets/8b92cb53-f872-404a-b76d-e3a3f77048ab)

Invalid message NACK:
![image](https://github.com/user-attachments/assets/ac4e0ebe-5db8-4811-9163-bcec4b47a34a)

**Comment by JohnNKing** [comment]
Connectivity should now be working. However, we need to verify that the results are being filed successfully.

To confirm: 
1. Is linefeed issue fixed?
2. Is the ISO-8859 vs UTF-8 issue working? (umol?)

**Comment by jorg3lopez** [comment]
Update:
- Invalid version ACK issue resolved. This was a bug on their side.
- [PR ](https://github.com/CDCgov/prime-reportstream/pull/15320/files)with linefeed issue fix has been merged. 
- We were able to send an ORU message to RS (staging) and the result message reached UCSD without any issues.
- Since we are dealing with other format related bugs (ISO-8859 vs UTF-8), we need more testing done in order to ensure that the messages are being delivered and UCSD is able to ingest them without any issues.
